File: algorithms/001.two_sum.py
# 001. Two Sum
# author: github.com/huklee
# O(n) TC solution using a complement map
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        targetMap = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in targetMap:
                j = targetMap[complement]
                return [min(i, j), max(i, j)] # Success
            else:
                targetMap[nums[i]] = i
        return [] # Fallback

File: algorithms/001.two-sum.go
func twoSum(nums []int, target int) []int {
    hashMap := make(map[int]int)
    for i, num := range nums{
        hashMap[num] = i
    }

    for i, num := range nums {
        complement := target - num
        if j, ok := hashMap[complement]; ok && i != j {
            return []int{i, j}
        }
    }
    return nil
}
